## NovaGen Research Labs

In [21]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,recall_score,classification_report
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import GradientBoostingClassifier , RandomForestClassifier , VotingClassifier

In [24]:
df=pd.read_csv("novagen_dataset.csv")


X=df.drop("Target",axis=1)
y=df["Target"]

X_train,X_test,y_train,y_test=train_test_split(
    X, y , test_size=0.2 , random_state=42,stratify=y
)


In [25]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Baseline Model-LogisticRegression

In [37]:
log_Reg=LogisticRegression(
    max_iter=1000,
    penalty="l2",
    solver="liblinear"

)

log_Reg.fit(X_train_scaled,y_train)

y_pred=log_Reg.predict(X_test_scaled)

print("Logistic Regression accuracy_score-",accuracy_score(y_test,y_pred)*100)
print("Logistic Regression recall_score-",recall_score(y_test,y_pred)*100)
print(classification_report(y_test,y_pred))

Logistic Regression accuracy_score- 81.41361256544502
Logistic Regression recall_score- 82.83132530120481
              precision    recall  f1-score   support

           0       0.81      0.80      0.80       914
           1       0.82      0.83      0.82       996

    accuracy                           0.81      1910
   macro avg       0.81      0.81      0.81      1910
weighted avg       0.81      0.81      0.81      1910



c:\Users\Lenovo\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\linear_model\_logistic.py:1135: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', and C=np.inf instead of penalty=None.
  warnings.warn(


## Model-2 KNN

In [35]:
knn=KNeighborsClassifier(
    n_neighbors=5,
    metric="euclidean"
)

knn.fit(X_train_scaled,y_train)

y_pred=knn.predict(X_test_scaled)

print("KNN accuracy_score-",accuracy_score(y_test,y_pred)*100)
print("KNN recall_score-",recall_score(y_test,y_pred)*100)
print(classification_report(y_test,y_pred))

KNN accuracy_score- 88.32460732984293
KNN recall_score- 88.35341365461848
              precision    recall  f1-score   support

           0       0.87      0.88      0.88       914
           1       0.89      0.88      0.89       996

    accuracy                           0.88      1910
   macro avg       0.88      0.88      0.88      1910
weighted avg       0.88      0.88      0.88      1910



## Model -3 Random Forest (Ensemble Learning)

In [38]:
rf = RandomForestClassifier(
    n_estimators=200,
    max_depth=None,
    random_state=42
)

rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))
print("Random Forest Recall:", recall_score(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_rf))


Random Forest Accuracy: 0.9382198952879581
Random Forest Recall: 0.9588353413654619
              precision    recall  f1-score   support

           0       0.95      0.92      0.93       914
           1       0.93      0.96      0.94       996

    accuracy                           0.94      1910
   macro avg       0.94      0.94      0.94      1910
weighted avg       0.94      0.94      0.94      1910



## Gradient Boosting Classifier

In [39]:
gb = GradientBoostingClassifier(
    n_estimators=150,
    learning_rate=0.1,
    max_depth=3,
    random_state=42
)

gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)

print("Gradient Boosting Accuracy:", accuracy_score(y_test, y_pred_gb))
print("Gradient Boosting Recall:", recall_score(y_test, y_pred_gb))
print(classification_report(y_test, y_pred_gb))


Gradient Boosting Accuracy: 0.9303664921465968
Gradient Boosting Recall: 0.9497991967871486
              precision    recall  f1-score   support

           0       0.94      0.91      0.93       914
           1       0.92      0.95      0.93       996

    accuracy                           0.93      1910
   macro avg       0.93      0.93      0.93      1910
weighted avg       0.93      0.93      0.93      1910



## Voting Classifier

In [40]:
vote=VotingClassifier(
    estimators=[
        ("lr",LogisticRegression(max_iter=1000,solver="liblinear")),
        ("knn",KNeighborsClassifier(n_neighbors=5)),
        ("rf",RandomForestClassifier(n_estimators=200,random_state=42))
    ],
    voting="soft"
)

vote.fit(X_train, y_train)

y_pred_gb = vote.predict(X_test)

print("Voting  Accuracy:", accuracy_score(y_test, y_pred_gb))
print("Voting  Recall:", recall_score(y_test, y_pred_gb))
print(classification_report(y_test, y_pred_gb))


Voting  Accuracy: 0.9047120418848168
Voting  Recall: 0.9156626506024096
              precision    recall  f1-score   support

           0       0.91      0.89      0.90       914
           1       0.90      0.92      0.91       996

    accuracy                           0.90      1910
   macro avg       0.90      0.90      0.90      1910
weighted avg       0.90      0.90      0.90      1910



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 82.8%  |
| KNN                  | 88.3%  |
| Random Forest        | 95.8%  |
| Gradient Boosting    | 94.9%  |
| Voting Classifier    | 93.07% |

### Best Classifier that we should use for NovaGen(based on Recall) - Random Forest with accuracy of 93.7%